source/opt/folding_rules.cpp

```diff
@@ -1463,90 +1464,115 @@ FoldingRule IntMultipleBy1() {
   };
 }
 
-FoldingRule CompositeConstructFeedingExtract() {
-  return [](IRContext* context, Instruction* inst,
-            const std::vector<const analysis::Constant*>&) {
-    // If the input to an OpCompositeExtract is an OpCompositeConstruct,
-    // then we can simply use the appropriate element in the construction.
-    assert(inst->opcode() == SpvOpCompositeExtract &&
-           "Wrong opcode.  Should be OpCompositeExtract.");
-    analysis::DefUseManager* def_use_mgr = context->get_def_use_mgr();
-    analysis::TypeManager* type_mgr = context->get_type_mgr();
+// Returns the number of elements that the |index|th in operand in |inst|
+// contributes to the result of |inst|.  |inst| must be an
+// OpCompositeConstructInstruction.
+uint32_t GetOperandElementCount(IRContext* context, const Instruction* inst,
+                                uint32_t index) {
+  assert(inst->opcode() == SpvOpCompositeConstruct);
+  analysis::DefUseManager* def_use_mgr = context->get_def_use_mgr();
+  analysis::TypeManager* type_mgr = context->get_type_mgr();
+
+  analysis::Vector* result_type =
+      type_mgr->GetType(inst->type_id())->AsVector();
+  if (result_type == nullptr) {
+    // If the result of the OpCompositeConstruct is not a vector then every
+    // operands corresponds to a single element in the result.
+    return 1;
+  }
 
-    // If there are no index operands, then this rule cannot do anything.
-    if (inst->NumInOperands() <= 1) {
-      return false;
-    }
+  // If the result type is a vector then the operands are either scalars or
+  // vectors. If it is a scalar, then it corresponds to a single element.  If it
+  // is a vector, then each element in the vector will be and element in the
+  // result.
+  uint32_t id = inst->GetSingleWordInOperand(index);
+  Instruction* def = def_use_mgr->GetDef(id);
+  analysis::Vector* type = type_mgr->GetType(def->type_id())->AsVector();
+  if (type == nullptr) {
+    return 1;
+  }
+  return type->element_count();
+}
 
-    uint32_t cid = inst->GetSingleWordInOperand(kExtractCompositeIdInIdx);
-    Instruction* cinst = def_use_mgr->GetDef(cid);
+// Returns the id and the literal index into the id from which the value of the
```

Comment:
By reading this description, it is a little bit difficult to know why the return type must be a vector and what is the meaning of `the id and the literal index into the id`.

It would be nice if the description explicitly mentioned that |inst| with a vector type returns `std::vector<Operand>` with **two elements** and it returns `std::vector<Operand>` with a single element for other cases.

Reply:
Yes, this is a very particular function.  I never got to a description and name that I was happy with.  I've changed it.  Let me know if this is better.

Reply:
It is better. It is fine to submit it.

Just consider it explains how it should be used specifically for `OpConstantExtract` not the function itself. One solution would be:
1. break it up in two parts i.e., vector and non-vector types
2. add a if-statement in the caller to determine if it is a vector type or not
3. call the two split functions for each case

Benefit: the description of each function would be more clear.